combine/combine.go

```diff
@@ -0,0 +1,306 @@
+	"strings"
+
+	"github.com/klauspost/pgzip"
+	"github.com/wtsi-ssg/wrstat/v3/dgut"
+)
+
+const bytesInMB = 1000000
+const pgzipWriterBlocksMultiplier = 2
+const dgutStoreBatchSize = 10000
+const dgutSumCols = 4
+const numSummaryColumnsDGUT = 3
+
+// ConcatenateAndCompress takes a list of open files as its input, and an open
+// file for its output. It compresses the output, and then stores there the
+// concatenated contents of the inputs.
+func ConcatenateAndCompress(inputs []*os.File, output *os.File) error {
+	compressor, closer, err := Compress(output)
+	if err != nil {
+		return err
+	}
+
+	err = Concatenate(inputs, compressor)
+	if err != nil {
+		return err
+	}
+
+	closer()
+
+	return nil
+}
+
+// Concatenate takes a list of open files as its input, and an io.Writer as its
+// output. It concatenates the contents of the inputs into the output.
+func Concatenate(inputs []*os.File, output io.Writer) error {
+	buf := make([]byte, bytesInMB)
+
+	for _, input := range inputs {
+		if _, err := io.CopyBuffer(output, input, buf); err != nil {
+			return err
+		}
+
+		if err := input.Close(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Compress takes an io writer as its input, and compresses it. It returns this,
+// along with a function to close the writer and an error status.
+func Compress(output io.Writer) (*pgzip.Writer, func(), error) {
+	zw := pgzip.NewWriter(output)
+
+	err := zw.SetConcurrency(bytesInMB, runtime.GOMAXPROCS(0)*pgzipWriterBlocksMultiplier)
+
+	return zw, func() {
+		err = zw.Close()
+		if err != nil {
+			return
+		}
+
+		if err != nil {
+			return
+		}
+	}, err
+}
+
+// Merger takes an input io.readCloser and an output io.writer, and defines how
+// we want to merge the content in the io.readCloser, and stream it to the
+// output io.writer.
+type Merger func(data io.ReadCloser, output io.Writer) error
+
+// Merge merges the inputs files and streams the content to the streamFunc.
+func Merge(inputs []*os.File, output io.Writer, streamFunc Merger) error {
+	inputFiles := make([]string, len(inputs))
+	for i, file := range inputs {
+		inputFiles[i] = file.Name()
+	}
+
+	sortMergeOutput, cleanup, err := MergeSortedFiles(inputFiles)
+	if err != nil {
+		return err
+	}
+
+	if err = streamFunc(sortMergeOutput, output); err != nil {
+		return err
+	}
+
+	if err = cleanup(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// MergeAndCompress takes a list of open files, an open output file, and a
+// Merger function to express the details of how the file contents should be
+// merged. It compresses the output, and stores the merged input contents in
+// there.
+func MergeAndCompress(inputs []*os.File, output *os.File, streamFunc Merger) error {
+	zw, closer, err := Compress(output)
+	if err != nil {
+		return err
+	}
+
+	err = Merge(inputs, zw, streamFunc)
+	if err != nil {
+		return err
+	}
+
+	closer()
+
+	return nil
+}
+
+// StatFiles finds and conatenates the stats files into the output, compressed.
```

Comment:
I'd expect the various file-type specific public methods to be in their own go file. Eg. StatFiles() should be in stat.go.